mmrazor/models/architectures/dynamic_op/base.py

```diff
@@ -1,25 +1,106 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from abc import ABC, abstractmethod
+from typing import Any, Optional, Set
 
 from torch import nn
 
-from mmrazor.models.mutables.mutable_channel import MutableChannel
+from mmrazor.models.mutables.base_mutable import BaseMutable
 
 
 class DynamicOP(ABC):
+    """Base class for dynamic OP. A dynamic OP usually consists of a normal
+    static OP and mutables, where mutables are used to control the searchable
+    (mutable) part of the dynamic OP.
+
+    Note:
+        When the dynamic OP has just been initialized, its forward propagation
+        logic should be the same as the corresponding static OP. Only after
+        the searchable part accepts the specific mutable through the
+        corresponding interface does the part really become dynamic.
+
+    Note:
+        All subclass should implement ``to_static_op`` API.
+
+    Args:
+        accepted_mutables (set): The string set of all accepted mutables.
+    """
+    accepted_mutables: Set[str] = set()
 
     @abstractmethod
     def to_static_op(self) -> nn.Module:
-        ...
+        """Convert dynamic OP to static OP.
+
+        Note:
+            The forward result for the same input between dynamic OP and its
+            corresponding static OP must be same.
+
+        Returns:
+            nn.Module: Corresponding static OP.
+        """
+
+    def check_if_mutables_fixed(self) -> None:
+        """Check if all mutables are fixed.
+
+        Raises:
+            RuntimeError: Error if a existing mutable is not fixed.
+        """
+
+        def check_fixed(mutable: Optional[BaseMutable]) -> None:
+            if mutable is not None and not mutable.is_fixed:
+                raise RuntimeError(f'Mutable {type(mutable)} is not fixed.')
+
+        for mutable in self.accepted_mutables:
+            check_fixed(getattr(self, f'{mutable}'))
+
+    @staticmethod
+    def get_current_choice(mutable: BaseMutable) -> Any:
+        """Get current choice of given mutable.
+
+        Args:
+            mutable (BaseMutable): Given mutable.
+
+        Raises:
+            RuntimeError: Error if `current_choice` is None.
+
+        Returns:
+            Any: Current choice of given mutable.
+        """
+        current_choice = mutable.current_choice
+        if current_choice is None:
+            raise RuntimeError(f'current choice of mutable {type(mutable)} '
+                               'can not be None at runtime')
+
+        return current_choice
 
 
```

Comment:
Suggestion:
    add unified interfaces, mutate_in_channels and mutate_out_channels, to ChannelDynamicOP.

Reply:
Preferably in each Dynamic OP, there may be `mutate_kernel_size` in `Conv`, `mutate_num_head` in `MultiHeadAttention`.
The variable of each module is xx, and the corresponding api is `mutate_xx`.

Reply:
My opinion is not  only leaving two apis mutate_in_channels and mutate_out_channels, just using the same api name to register mutable channel, rather than 'mutate_in_channels' in dyanmicconv and 'mutate_num_features' in dynamicbn.